File: src/go/day7/puzzle2.go
package main

import (
	"math"
)

func isSolveableWithConcat(eq equation) bool {
	if eq.target < 0 {
		return false
	}
	if len(eq.terms) == 1 {
		return eq.target == eq.terms[0]
	}
	last := eq.terms[len(eq.terms)-1]
	// divide target by last
	if eq.target%last == 0 && isSolveableWithConcat(equation{eq.target / last, eq.terms[:len(eq.terms)-1]}) {
		return true
	}

	// check unconcat
	// x1, x2 := unconcat(eq.target, last), _unconcat(eq.target, last)
	// if x1 != x2 {
	// 	panic(fmt.Errorf("unconcat(%d, %d) = %d, expected %d", eq.target, last, x1, x2))
	// }
	if unconcat := unconcat(eq.target, last); unconcat < eq.target && isSolveableWithConcat(equation{unconcat, eq.terms[:len(eq.terms)-1]}) {
		return true
	}

	// subtract last target
	if isSolveableWithConcat(equation{eq.target - last, eq.terms[:len(eq.terms)-1]}) {
		return true
	}
	return false
}

func unconcat(a, b int) int {
	if a == b {
		return 0
	}
	order := math.Ceil(math.Log10(float64(b) + 0.1))
	exp := int(math.Pow10(int(order)))
	if a%exp == b {
		return (a - b) / exp
	}
	return a
}

File: src/python/day7/puzzle2.py
from .equation import equation
from math import log10, ceil, pow

def isSolveableWithConcat(target, terms):
    return _isSolveableWithConcat(target, terms, len(terms)-1)

def _isSolveableWithConcat(target, terms, head):
    if target < 0:
        return False

    if head == 0:
        return target == terms[head]

    last = terms[head]

    # try division
    if target % last == 0 and _isSolveableWithConcat(target//last, terms, head-1):
        return True

    # try deconcatenating
    unconcatenated = unconcat(target, last)
    if unconcatenated < target and _isSolveableWithConcat(unconcatenated, terms, head-1):
        return True

    # try subtraction
    return _isSolveableWithConcat(target-last, terms, head-1)

def unconcat(a, b):
    if a == b:
        return 0

    order = int(ceil(log10(b+0.1)))
    exp = int(pow(10, order))
    if a%exp == b:
        return (a-b)/exp
    return a

if __name__ == "__main__":
    print(isSolveableWithConcat(10, [1, 9]))
    print(isSolveableWithConcat(10, [1, 8]))
    print(isSolveableWithConcat(10, [1, 10]))
    print(isSolveableWithConcat(10, [3, 3, 1]))